### ISSUE TITLE
Support asynchronous pipeline for cluster 

### ISSUE THREAD

**Issue by thomas-burke** [redis-rs/redis-rs#1348]
Hi All, 
I'm unable to pipeline with cluster_async as expected. 

The best way to observe this issue is to duplicate this line 
    https://github.com/redis-rs/redis-rs/blob/main/redis/tests/test_cluster_async.rs#L338
and change it to use a different key name like this:
```
                pipe.add_command(cmd("SET").arg("test").arg("test_data").clone());
                pipe.add_command(cmd("SET").arg("best").arg("test_data").clone());
```
Note that `"test"` and `"best"` map to different SLOTs

After the change above I get this error when I run the test:
```
test cluster_async::test_async_cluster_basic_pipe::case_1_tokio ... thread 'cluster_async::test_async_cluster_basic_pipe::case_1_tokio' panicked at redis/tests/test_cluster_async.rs:350:10:
called `Result::unwrap()` on an `Err` value: Received crossed slots in pipeline- CrossSlot
```

TLDR; I can pipeline when I use synchronous `cluster`. 
> The synchronous implementation and cluster_pipe is OK. The code below sets both keys without a CrossSlot Err
> 
> Example:
> ```
>     let nodes = vec!["redis://127.0.0.1:6379/","redis://127.0.0.1:6380/","redis://127.0.0.1:6381/"];
>     let client = ClusterClient::new(nodes).unwrap();
>     let mut binding = cluster_pipe();
>     let pipe_s = binding.set("test", "ABC").ignore().set("best", "123").ignore();
>     let mut connection_s = client.get_connection().unwrap();
> 
>     let result_s = pipe_s.exec(&mut connection_s);
>     println!("Result from synchronous redis cluster pipeline exec {result_s:?}");
> 
>     let rv_s:redis::Value = connection_s.get("test").unwrap();
>     println!("Value1 from redis cluster (pipelined commands) {rv_s:?}");
>     let rv_s:redis::Value = connection_s.get("best").unwrap();
>     println!("Value2 from redis cluster (pipelined commands) {rv_s:?}");
> ```

I looked for an issue related to this but the closest issue I could find was https://github.com/redis-rs/redis-rs/issues/1255 Please let me know if this is a duplicate of 1255 or some other issue.

**Comment by nihohit** [comment]
I dislike this incongruity between the sync and async client, but IMO this is the correct behavior for the async client*, and there's a logic guiding this difference.

The sync client can only execute a single command/pipeline at a time - there's no concurrency there, as far as the external API allows. This means that pipelining is critical for reasonable performance, and a pipeline could be taken apart to its component commands, and sent on each connection separately. But, the question is *how* do you split a pipeline - if you have GET, CLUSTER FAILOVER, GET commands in the pipeline, which node should receive the CLUSTER FAILOVER command? the first or the second? And also, you can't perform a transaction this way - an atomic pipeline is meaningless if split apart.

The async client allows for concurrent requests - just clone the client, pass it to a different thread / task, and the client will internally multiplex commands without the need for pipelines, This means that the cost of not pipelining commands is much smaller, while the danger of incorrect behavior is greater - now if you split a pipeline apart, not only is it unclear which node should receive the in-between commands, also the order of commands between concurrent pipelines is uncertain, and we can't ensure that allowing this won't be a footgun. 
The greater correctness danger, combined with the smaller performance win, led us to enforce CROSSLOTS on pipelines in the async client.

but, these issues are relevant only for pipelines. If you just want to group your SETs or GETs, you can use the MSET/MGET commands with an async cluster connection - as single commands, the async cluster can take them apart and send them to separate nodes, because the ordering & correctness issues are significantly smaller for single commands. Just don't use them in a pipeline :)

* of course, I wrote it, so it's not surprising :)
